src/cluster/placement/algo/mirrored.go

```diff
@@ -288,17 +612,34 @@ func allInstancesInState(
 	instanceIDs map[string]struct{},
 	p placement.Placement,
 	forEachShardFn func(s shard.Shard) bool,
+	exactly bool,
```

Comment:
This additional logic makes this function more complex. Maybe define a new function `instancesInState(...)`?